src/Recurr/Transformer/InstanceGenerator.php

```diff
@@ -0,0 +1,771 @@
+
+            if (empty($rDates)) {
+                // No next date could be found, so we're done here.
+                break;
+            }
+
+            // Get the lowest of all the `$r_dates` This is the "next" instance date for the series, so long as
+            // it's not also found in with the EXDATES or the EXRULEs.
+            $minRDate = $this->minDate($rDates);
+
+            $dateIsExcluded = false;
+
+            $exGeneratorsToKeep = [];
+
+            foreach ($exGenerators as $generator) {
+                if (!$generator->valid()) {
+                    // This generator has reached it's end... Skip it and don't keep it.
+                    continue;
+                }
+
+                // Iterate through all exdates until we reach the `$min_r_date`
+                while ($generator->valid() && $generator->current()->getTimestamp() < $minRDate->getTimestamp()) {
+                    $generator->next();
+                }
+
+                // If the current EXDATE matches the current `$min_r_date`, then the date should be excluded from the
+                // series.
+                if ($generator->valid() && $generator->current()->getTimestamp() === $minRDate->getTimestamp()) {
+                    $dateIsExcluded = true;
+                }
+
+                $exGeneratorsToKeep[] = $generator;
+            }
+
+            $exGenerators = $exGeneratorsToKeep;
+
+            if (!$dateIsExcluded) {
+                $iterations++;
+                yield $minRDate;
+            }
+
+            foreach ($rGenerators as $generator) {
+                // Advance all the `$r_generators` that yielded the previous result.
+                if ($generator->current()->getTimestamp() === $minRDate->getTimestamp()) {
+                    $generator->next();
+                }
+            }
+            // Keep looping so long as the `$r_generators` are yielding dates and we're under the iteration limit.
+        } while (!empty($rGenerators) && (null === $iterationLimit || $iterations < $iterationLimit));
+
+    }
+
+    /**
+     * Returns a generator that provides a given array of dates, copied with the given time zone.
+     *
+     * @param DateTime[] $datetimes The dates to be yielded by the generator.
+     * @param DateTimeZone $timeZone The timezone context to give all the yielded dates.
+     * @return Generator A generator that yields {@link DateTime} values representing instance start time.
+     */
+    private function generateFromDates(array $datetimes, DateTimeZone $timeZone)
+    {
+        $compare = function (DateTime $datetime1, DateTime $datetime2) {
+            return ($datetime1->getTimestamp() - $datetime2->getTimestamp());
+        };
+
+        usort($datetimes, $compare);
+
+        foreach ($datetimes as $date) {
+            $result = clone $date;
+            $result->setTimezone($timeZone);
+            $result->getTimestamp(); // Fixes PHP timezone bug.
```

Comment:
Probably a good idea to link it here, since it's undocumented in this project or the upstream.